### ISSUE TITLE
Contains widget on "N2" strain page has too many clones to handle with standard display

### ISSUE THREAD

**Issue by chris-grove** [WormBase/website#3712]
If I open the "N2" strain page on WormBase with the "Contains" widget open, it has, of course, lots of associated clone objects (15,355). If I click on the clones list to expand and view, the page basically becomes non-operational and unresponsive forcing me to close the page or refresh. Can we have a another way to display these clones, either as a download option or in a table format so that the page doesn't crash when browsing the data?

**Comment by tuli** [comment]
@chris-grove  I'm not on the web team (obviously!), but have dealt with similar issues with Variation and Gene objects having large number of Contains objects (#2809 #3583).

Do you think users really expect to go to N2 strain page and see all 150,000 clones? Maybe we'd be better pointing them to the elegans Species pages.

**Comment by sibyl229** [comment]
Hi @chris-grove the temporary solution that's used elsewhere (when when there's a large number of objects) is to display a link to search page. In this case, it can show a count, a link to http://staging.wormbase.org/search/clone/N2, and a link to FTP site, where they can download and extract information they need. 

Search has its own problems - false positives and false negatives. So it's not a good solution yet. But to get this many objects directly from ACe will be extremely slow.

**Comment by chris-grove** [comment]
@sibyl229  Yeah, I was thinking something like that, similar to what we do for RNAi objects/experiments on phenotype pages where we have large numbers (1000s) of RNAi experiments, like:

http://www.wormbase.org/species/all/phenotype/WBPhenotype:0000050#04--10

@tuli I imagine we can do the same for the variations and such

**Comment by sibyl229** [comment]
Hi @chris-grove The Contains widget is working for N2 strain now: http://staging.wormbase.org/species/c_elegans/strain/N2#01--10 But I noticed a separate problem in the Overview widget, the Genotype field goes through all the clones, and tries to replace part of the genotype text with links to the clones. Is it supposed to do that? Thanks

**Comment by chris-grove** [comment]
@sibyl229  OK thanks, the "Contains" widget looks good (i.e. is functional). The remaining issue is that the "Clone: 15355 found" hyperlink to the general search page indicates that there are "Too many results to download. Please use our FTP site."

![screen shot 2015-04-13 at 2 19 37 pm](https://cloud.githubusercontent.com/assets/611382/7125737/4a164bbc-e1e8-11e4-81c6-169a84fe697c.png)

and I gather that this is the default behavior for any searches with ~>100K results. The problem is that I have no idea where to look on the FTP site to find this data. I suppose there is no real solution to this and users will just have to explore; a bigger issue is that I don't even know if this data actually does exist on the FTP site. Is there any way to point users to the correct location?

As for the Genotype field in the Overview widget, I really don't know, although it certainly doesn't sound right to me.

**Comment by tuli** [comment]
@sibyl229 @chris-grove 
"But I noticed a separate problem in the Overview widget, the Genotype field goes through all the clones, and tries to replace part of the genotype text with links to the clones. Is it supposed to do that? Thanks"

No, that sounds very wrong. 

Regarding the Contains widget. I also don't know if this info is available on the FTP site and if so, where, but I did suggest we simply disable the "15355 clones found" link and change it to "15355 clones found. Further clone information may be found on the Species page" where Species is a link to http://www.wormbase.org/species/c_elegans/clone 

I have never come across anyone wanting to download all these clones.

Incidentally, the Browse widget on http://www.wormbase.org/species/c_elegans/clone has download options and I was successfully able to download and view the clones in cvs format.

**Comment by chris-grove** [comment]
Found another strain page where the "Contains" widget throws an error, presumably due to the 6,816 variations it contains (see Tree Display):

https://wormbase.org/species/c_elegans/strain/MY15#015--10

![screen shot 2018-07-25 at 10 02 23 am](https://user-images.githubusercontent.com/611382/43205855-515d1e12-8ff2-11e8-903f-dba7a6420375.png)

**Comment by sibyl229** [comment]
Hi @chris-grove I just cached the content for the "contains" widget for MY15. It should be able to load, even though a bit slowly.

Hi @tharris can I confirm with you the slow d2c endpoints that needs precaching? Here is my list: https://github.com/WormBase/website/blob/staging/lib/WormBase/Web/Controller/REST.pm#L1092-L1104

**Comment by chris-grove** [comment]
@sibyl229  OK thanks, the widget is loading now (albeit slowly as you said, which is fine (better than throwing an error message)).

The original N2 clone issue is still problematic, so I will leave this ticket open. The problem now isn't that the N2 page "Contains" widget doesn't load, it does:

https://wormbase.org/species/c_elegans/strain/N2#015--10

but it only reads "15355 found" which is hyperlinked but leads to a null search results page, apparently trying to search the ?Clone class for the string "N2".
